ipam/allocator.go

```diff
@@ -44,71 +40,111 @@ type Allocator struct {
 func NewAllocator(lcDs, glDs datastore.DataStore) (*Allocator, error) {
 	a := &Allocator{}
 
-	a.localSubnets = &PoolsConfig{
-		subnets: map[SubnetKey]*PoolData{},
-		id:      dsConfigKey + "/Pools",
-		scope:   datastore.LocalScope,
-		ds:      lcDs,
-		alloc:   a,
-	}
-
-	a.globalSubnets = &PoolsConfig{
-		subnets: map[SubnetKey]*PoolData{},
-		id:      dsConfigKey + "/Pools",
-		scope:   datastore.GlobalScope,
-		ds:      glDs,
-		alloc:   a,
-	}
-
+	// Load predefined subnet pools
 	a.predefined = map[string][]*net.IPNet{
 		localAddressSpace:  initLocalPredefinedPools(),
 		globalAddressSpace: initGlobalPredefinedPools(),
 	}
 
-	a.addrSpace2Configs = map[string]*PoolsConfig{
-		localAddressSpace:  a.localSubnets,
-		globalAddressSpace: a.globalSubnets,
+	// Initialize bitseq map
+	a.addresses = make(map[SubnetKey]*bitseq.Handle)
+
+	// Initialize address spaces and reconcile with store if necessary
+	if err := a.initAddressSpaces([]datastore.DataStore{lcDs, glDs}); err != nil {
+		return nil, err
 	}
 
-	a.addresses = make(map[SubnetKey]*bitseq.Handle)
+	return a, nil
+}
 
-	cfgs := []struct {
-		cfg *PoolsConfig
-		dsc string
-	}{
-		{a.localSubnets, "local"},
-		{a.globalSubnets, "global"},
+func scopeToAs(scope string) string {
+	switch scope {
+	case datastore.LocalScope:
+		return localAddressSpace
+	case datastore.GlobalScope:
+		return globalAddressSpace
 	}
-	// Get the initial local/global pools configfrom the datastores
-	var inserterList []func() error
-	for _, e := range cfgs {
-		if e.cfg.ds == nil {
+
+	return ""
+}
+
+func asToScope(as string) string {
+	switch as {
+	case localAddressSpace:
+		return datastore.LocalScope
+	case globalAddressSpace:
+		return datastore.GlobalScope
+	}
+
+	return ""
+}
+
+func (a *Allocator) initAddressSpaces(stores []datastore.DataStore) error {
```

Comment:
If you do not see a need to call this function other than inside the Allocator constructor, then you could pass as parameter the scope along each datastore, which is known by the caller.

This would avoid the need for the extra function `scopetoAs()` and also `asToScope()` which is actually not being used anywhere.